src/curve.rs

```diff
@@ -10,81 +12,97 @@ pub struct Curve<F: Field> {
   two:   F,
 }
 
-pub struct CurvePoint<F: Field> {
-  pub curve: Curve<F>,
-  pub point: PointOrInfinity<F>,
-}
+pub trait CurveParams: 'static + Copy + Clone + fmt::Debug + Default + Eq + Ord {
+  /// Integer field element type
+  type FieldElement: Field;
+  /// Order of this elliptic curve, i.e. number of elements in the scalar field.
+  const ORDER: u32;
+  /// Coefficient `a` in the Weierstrass equation of this elliptic curve.
+  const EQUATION_A: Self::FieldElement;
+  /// Coefficient `b` in the Weierstrass equation of this elliptic curve.
+  const EQUATION_B: Self::FieldElement;
+  /// Generator of this elliptic curve.
+  const GENERATOR: (Self::FieldElement, Self::FieldElement);
+  // hack: 3 and 2 to satisfy the Add<AffinePoint> trait implementation
+  const THREE: Self::FieldElement;
+  const TWO: Self::FieldElement;
 
-#[derive(Clone, Copy)]
-pub struct Point<F: Field> {
-  x: F,
-  y: F,
+  // maybe Curve::uint type is diff from PCP::fieldelement type
+  // maybe:
+  // type AffinePoint;
+  // type ProjectivePoint;
+  // type Scalar;
 }
 
-#[derive(Clone, Copy)]
-pub enum PointOrInfinity<F: Field> {
-  Point(Point<F>),
-  Infinity,
+/// The Elliptic curve $y^2=x^3+3$, i.e.
+/// - a = 0
+/// - b = 3
+/// - generator todo
+/// - order todo
+/// - field element type todo, but mock with u64 - bad thor, u64 does not implement p3_field
+#[derive(Copy, Clone, Debug, Default, Eq, PartialEq, PartialOrd, Ord)]
+pub struct C101;
+
+impl CurveParams for C101 {
+  type FieldElement = crate::field::PlutoField;
+
+  const EQUATION_A: Self::FieldElement = PlutoField::const_new(0);
+  const EQUATION_B: Self::FieldElement = PlutoField::const_new(3);
+  const GENERATOR: (Self::FieldElement, Self::FieldElement) = todo!();
+  const ORDER: u32 = PlutoField::ORDER_U32;
+  const THREE: Self::FieldElement = PlutoField::const_new(3);
+  const TWO: Self::FieldElement = PlutoField::const_new(2);
 }
 
-impl<F: Field> Curve<F> {
-  pub fn new(a: F, b: F) -> Self {
-    Self {
-      a,
-      b,
-      three: <F as AbstractField>::from_canonical_u8(3),
-      two: <F as AbstractField>::from_canonical_u8(2),
-    }
-  }
+/// An Affine Coordinate Point on a Weierstrass elliptic curve
+#[derive(Clone, Debug, Copy)]
+pub struct AffinePoint<C: CurveParams> {
+  x:     C::FieldElement,
+  y:     C::FieldElement,
+  /// is the point the point at infinity
+  infty: bool,
```

Comment:
I think this should be an enum still. E.g., something like:
```suggestion
pub enum AffinePoint<C: CurveParams> {
  Point(C::FieldElement, C::FieldElement),
  Infinity,
```

Reply:
This makes sense as a point in ONLY one of these two, which makes it feel more type theoretically correct. But it also might be a little more verbose. Could go either way.